src/libYARP_OS/src/HierarchicalMutex.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2006-2018 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * BSD-3-Clause license. See the accompanying LICENSE file for details.
+ */
+
+#include <yarp/os/HierarchicalMutex.h>
+#include <climits>
+
+using yarp::os::HierarchicalMutex;
+
+class HierarchicalMutex::Private
+{
+    std::mutex internal_mutex;
+    unsigned long const hierarchy_value;
+    unsigned long previous_hierarchy_value;
+    static thread_local unsigned long this_thread_hierarchy_value;
+
+    void check_for_hierarchy_violation()
+    {
+        if(this_thread_hierarchy_value <= hierarchy_value)
+        {
+            throw std::logic_error("mutex hierarchy violated");
+        }
+    }
+    void update_hierarchy_value()
```

Comment:
For readability, I would separete method implementations with at least one empty line. Pleas ealso check other methods of this PR.